### PR TITLE
add NUTS kernel

### PR DESCRIPTION
**Pull request by rlouf**
In this PR we add the NUTS kernel.

### CODE REVIEW

File: aehmc/proposals.py
Comment (rlouf) [reply aehmc/proposals.py:51]: Good to know, I must have been looking in the wrong version of the documentation.

Comment (ricardoV94) [reply aehmc/proposals.py:51]: I am not sure whether it shows up in the documentation actually :/

File: aehmc/termination.py
Comment (rlouf) [inline aehmc/termination.py:48]: @brandonwillard could you help me understand the following error?

<details><summary>Error message</summary>
<p>

```
E           ValueError: Dimension 0 in Rebroadcast's input was supposed to be 1 (got 3 instead)
E           Apply node that caused the error: Rebroadcast{1}(TensorConstant{(3,) of 1.0})
E           Toposort index: 4
E           Inputs types: [TensorType(float64, vector)]
E           Inputs shapes: [(3,)]
E           Inputs strides: [(8,)]
E           Inputs values: [array([1., 1., 1.])]
E           Outputs clients: [[IncSubtensor{InplaceSet;int32}(Alloc.0, Rebroadcast{1}.0, ScalarFromTensor.0)]]
E           
E           Backtrace when the node is created (use Aesara flag traceback__limit=N to make it longer):
E             File "/home/remi/.virtualenvs/aehmc/lib/python3.9/site-packages/_pytest/python.py", line 1641, in runtest
E               self.ihook.pytest_pyfunc_call(pyfuncitem=self)
E             File "/home/remi/.virtualenvs/aehmc/lib/python3.9/site-packages/pluggy/hooks.py", line 286, in __call__
E               return self._hookexec(self, self.get_hookimpls(), kwargs)
E             File "/home/remi/.virtualenvs/aehmc/lib/python3.9/site-packages/pluggy/manager.py", line 93, in _hookexec
E               return self._inner_hookexec(hook, methods, kwargs)
E             File "/home/remi/.virtualenvs/aehmc/lib/python3.9/site-packages/pluggy/manager.py", line 84, in <lambda>
E               self._inner_hookexec = lambda hook, methods, kwargs: hook.multicall(
E             File "/home/remi/.virtualenvs/aehmc/lib/python3.9/site-packages/pluggy/callers.py", line 187, in _multicall
E               res = hook_impl.function(*args)
E             File "/home/remi/.virtualenvs/aehmc/lib/python3.9/site-packages/_pytest/python.py", line 183, in pytest_pyfunc_call
E               result = testfunction(**testargs)
E             File "/home/remi/projects/ampersand/aehmc/tests/test_termination.py", line 52, in test_termination_update
E               updated = update(termination_state, momentum_sum, momentum, step)
E             File "/home/remi/projects/ampersand/aehmc/aehmc/termination.py", line 46, in update
E               aet.set_subtensor(momentum_ckpt[idx_max], momentum),
```

</p>
</details>

Comment (brandonwillard) [reply aehmc/termination.py:48]: I don't know what's generating that `Rebroadcast`, so I would try to determine that first.  The `set_subtensor` (and/or the `*Subtensor*` implied by `momentum_ckpt[idx_max]`) is apparently involved, so try printing that sub-graph with `aesara.dprint`.

Otherwise, it looks like the shapes/dimension of the inputs are incompatible with the graph, because `Rebroadcast{1}` is essentially a check that makes sure the first dimension is broadcastable (i.e. its size is equal to 1), but it has been given an array with a dimension that isn't `1` (i.e. `np.array([1., 1., 1.])`).

**Comment by rlouf** [comment]
@brandonwillard it looks like I still have a problem with a `RandomStream` node that I do not understand. The stack trace I guet when executing `test_dynamic_expansion` in `test_trajectory.py` seems to indicates that the problems lies comes from the dynamic integration. However the test `test_dynamic_integration_divergence` does pass. The problem would appear when I share the `RandomStream` between two functions. Is there a workaround?

<details>
    <summary>Stack trace</summary>
```
WARNING (aesara.tensor.blas): Using NumPy C-API based implementation for BLAS functions.
Traceback (most recent call last):
  File "/home/remi/projects/ampersand/aehmc/example.py", line 64, in <module>
    fn = aesara.function((), result)
  File "/home/remi/.virtualenvs/aehmc/lib/python3.9/site-packages/aesara/compile/function/__init__.py", line 337, in function
    fn = pfunc(
  File "/home/remi/.virtualenvs/aehmc/lib/python3.9/site-packages/aesara/compile/function/pfunc.py", line 524, in pfunc
    return orig_function(
  File "/home/remi/.virtualenvs/aehmc/lib/python3.9/site-packages/aesara/compile/function/types.py", line 1972, in orig_function
    m = Maker(
  File "/home/remi/.virtualenvs/aehmc/lib/python3.9/site-packages/aesara/compile/function/types.py", line 1586, in __init__
    fgraph, additional_outputs = std_fgraph(inputs, outputs, accept_inplace)
  File "/home/remi/.virtualenvs/aehmc/lib/python3.9/site-packages/aesara/compile/function/types.py", line 190, in std_fgraph
    fgraph = FunctionGraph(orig_inputs, orig_outputs, update_mapping=update_mapping)
  File "/home/remi/.virtualenvs/aehmc/lib/python3.9/site-packages/aesara/graph/fg.py", line 167, in __init__
    self.import_var(output, reason="init")
  File "/home/remi/.virtualenvs/aehmc/lib/python3.9/site-packages/aesara/graph/fg.py", line 337, in import_var
    self.import_node(var.owner, reason=reason, import_missing=import_missing)
  File "/home/remi/.virtualenvs/aehmc/lib/python3.9/site-packages/aesara/graph/fg.py", line 402, in import_node
    raise MissingInputError(error_msg, variable=var)
aesara.graph.fg.MissingInputError: Input 1 (<TensorType(float64, vector)>) of the graph (indices start from 0), used to compute Elemwise{mul,no_inplace}(Inplac
eDimShuffle{x}.0, <TensorType(float64, vector)>), was not provided and not given a value. Use the Aesara flag exception_verbosity='high', for more information
on this error.

Backtrace when that variable is created:

  File "/home/remi/projects/ampersand/aehmc/example.py", line 63, in <module>
    result = expand(proposal, state, state, state[1], termination_state, energy)
  File "/home/remi/projects/ampersand/aehmc/aehmc/trajectory.py", line 428, in expand
    ), updates = trajectory_integrator(
  File "/home/remi/projects/ampersand/aehmc/aehmc/trajectory.py", line 236, in integrate
    traj, updates = aesara.scan(
```
</details>

The following simplified example does work as expected:

```python
from typing import Callable

import aesara
import aesara.tensor as aet
from aesara.tensor.random.utils import RandomStream

rng = RandomStream(59)


def bernoulli_adder(srng: RandomStream):
    def add_bernoulli(b):
        def one_step(x):
            a = aet.where(srng.bernoulli(0.5), 1, -1)
            c = a + x
            return c

        res, _ = aesara.scan(one_step, outputs_info=(b,), n_steps=10)
        return res

    return add_bernoulli


def normal_adder(srng: RandomStream, add_bernoulli: Callable):
    def add_normal(init):
        def one_step(a):
            res = add_bernoulli(a)
            final = res[-1]
            norm = srng.normal(0, 1)
            return final + norm

        res, _ = aesara.scan(one_step, outputs_info=(init,), n_steps=10)
        return res

    return add_normal


init = aet.scalar("init")
add_bernoulli = bernoulli_adder(rng)
add_normal = normal_adder(rng, add_bernoulli)
result = add_normal(init)

fn = aesara.function((init,), result)

print(fn(10))
```

Which makes me think that the issue may be more trivial. But at this point I need another pair of eyes to look at the code.

**Comment by brandonwillard** [comment]
I don't see a `test_dynamic_expansion` function.

**Comment by rlouf** [comment]
`test_multiplicative_expansion`, sorry.

**Comment by brandonwillard** [comment]
The missing input variable appears to be introduced by/within [`new_potential_energy_grad`](https://github.com/aesara-devs/aehmc/blob/711a2d6fa22c004f6720d490fbb33d72fd7941da/aehmc/integrators.py#L58).  If you set a name for that (e.g. `new_potential_energy_grad.name = "new_potential_energy_grad"`) the error will say that the missing variable is `new_potential_energy_grad[t-1]`.

**Comment by brandonwillard** [comment]
OK, I think the issue might have to do with the [`outputs_info` arguments to the `scan` here](https://github.com/aesara-devs/aehmc/blob/711a2d6fa22c004f6720d490fbb33d72fd7941da/aehmc/trajectory.py#L236).  It looks like quite a few arguments are duplicated&mdash;including the `new_potential_energy_grad` variable in the error message.  `Scan` might not be handling duplicate arguments well.

**Review by brandonwillard — CHANGES_REQUESTED** [review]
It would help a lot if we used `nametuple`s instead of tuple indexing (e.g. for the `sample_proposal` variables, `traj`, etc.)

**Comment by rlouf** [comment]
@brandonwillard it seems like we don't have any problems with the updates inside the `scan` operators. I fixed a bunch of type mismatch errors, and I am now stuck with the following error:

```
TypeError: ('The following error happened while compiling the node', do_whileall_inplace,cpu,scan_fn}(TensorConstant{10}, IncSubtensor{InplaceSet;:int64:}.0, IncSubtensor{InplaceSet;:int64:}.0, IncSubtensor{InplaceSet;:int64:}.0, IncSubtensor{InplaceSet;:int64:}.0, DeepCopyOp.0, IncSubtensor{Set;:int64:}.0, DeepCopyOp.0, IncSubtensor{Set;:int64:}.0, DeepCopyOp.0, DeepCopyOp.0, DeepCopyOp.0, DeepCopyOp.0, DeepCopyOp.0, DeepCopyOp.0, DeepCopyOp.0, DeepCopyOp.0, DeepCopyOp.0, IncSubtensor{InplaceSet;:int64:}.0, DeepCopyOp.0, IncSubtensor{InplaceSet;:int64:}.0, DeepCopyOp.0, RandomGeneratorSharedVariable(<Generator(PCG64) at 0x7F709EB79D60>), RandomGeneratorSharedVariable(<Generator(PCG64) at 0x7F70BC80F3C0>), RandomGeneratorSharedVariable(<Generator(PCG64) at 0x7F709C9FE9E0>), RandomGeneratorSharedVariable(<Generator(PCG64) at 0x7F709C6BF9E0>)), '\n', "Inconsistency in the inner graph of scan 'scan_fn' : an input and an output are associated with the same recurrent state and should have the same type but have type 'TensorType(float64, (True,))' and 'TensorType(float64, vector)' respectively.")
```

As you can see in the `test_trajectory.py` file I did print the inputs to `expand`, and the output of `expand_once` (commented out). Inputs and outputs have the same shape. I don't really know what else to try.

**Comment by brandonwillard** [comment]
> @brandonwillard it seems like we don't have any problems with the updates inside the `scan` operators. I fixed a bunch of type mismatch errors, and I am now stuck with the following error:
> 
> ```
> TypeError: ('The following error happened while compiling the node', do_whileall_inplace,cpu,scan_fn}(TensorConstant{10}, IncSubtensor{InplaceSet;:int64:}.0, IncSubtensor{InplaceSet;:int64:}.0, IncSubtensor{InplaceSet;:int64:}.0, IncSubtensor{InplaceSet;:int64:}.0, DeepCopyOp.0, IncSubtensor{Set;:int64:}.0, DeepCopyOp.0, IncSubtensor{Set;:int64:}.0, DeepCopyOp.0, DeepCopyOp.0, DeepCopyOp.0, DeepCopyOp.0, DeepCopyOp.0, DeepCopyOp.0, DeepCopyOp.0, DeepCopyOp.0, DeepCopyOp.0, IncSubtensor{InplaceSet;:int64:}.0, DeepCopyOp.0, IncSubtensor{InplaceSet;:int64:}.0, DeepCopyOp.0, RandomGeneratorSharedVariable(<Generator(PCG64) at 0x7F709EB79D60>), RandomGeneratorSharedVariable(<Generator(PCG64) at 0x7F70BC80F3C0>), RandomGeneratorSharedVariable(<Generator(PCG64) at 0x7F709C9FE9E0>), RandomGeneratorSharedVariable(<Generator(PCG64) at 0x7F709C6BF9E0>)), '\n', "Inconsistency in the inner graph of scan 'scan_fn' : an input and an output are associated with the same recurrent state and should have the same type but have type 'TensorType(float64, (True,))' and 'TensorType(float64, vector)' respectively.")
> ```
> 
> As you can see in the `test_trajectory.py` file I did print the inputs to `expand`, and the output of `expand_once` (commented out). Inputs and outputs have the same shape. I don't really know what else to try.


OK, I'll take a look at it now; otherwise, that's an awful error message!  We need to fix that in Aesara.

**Comment by brandonwillard** [comment]
A note about that kind of error:

Whenever you see an error involving a discrepancy between two symbolic variables like `TensorType(float64, (True,))` and `TensorType(float64, vector)`, it's often because to the former variable is an "over-specialized" instance of the latter.

For example, if we define a generic symbolic vector `x = at.vector()`, it will have an `x.type` of `TensorType(float64, vector)`.  If we create a very specific instance of a vector via `at.as_tensor(np.r_[1.0])`, we'll get a `y.type` of `TensorType(float64, (True,))`.  

Both `x` and `y` are vectors, but the irreconcilable difference between the two types is the broadcast patterns (i.e. the dimensions that are known to be one).  In other words, `x.type.broadcastable == [False]` (i.e. the `vector` part in the `str` output of `x.type`) and `y.type.broadcastable == [True]` (i.e. the `(True,)` part in the `str` output of `y.type`).

`aesara.tensor.as_tensor[_variable]` is sometimes the cause of such errors&mdash;as the example above illustrates how the discrepancy can arise from its use&mdash;but it can also be due to an `Op` that does a poor job of inferring the broadcast pattern of its output(s), or a number of other related things.

**Comment by brandonwillard** [comment]
FYI: If we stop in the debugger at that error, we can see exactly where one of the offending variables (specifically, the latter `TensorType(float64, vector)`, which is one of the `Scan`'s "inner-output" variables) was created:
```python
pdb> pp self.outputs[inner_oidx].tag.trace[-1][-1]
('.../aehmc/aehmc/trajectory.py',
 463,
 'where_state',
 'q = aet.where(do_pick_left, q_left, q_right)')
```

File: aehmc/trajectory.py
Comment (brandonwillard) [inline aehmc/trajectory.py:464]: Are these supposed to be vectorized conditions, or scalar conditions that choose one vector or another?  

In the tests, `do_pick_left` is a scalar and `q_left` and `q_right` are vectors (albeit vectors of length one).  If you make `q_left` and `q_right` scalars, this should work like an `if`-`else`; otherwise, if `q_left` and `q_right` are not necessarily scalars, you might be better off using `ifelse`.

Comment (rlouf) [reply aehmc/trajectory.py:464]: I need to use `ifelse` instead of `where`.

File: aehmc/proposals.py
Comment (brandonwillard) [inline aehmc/proposals.py:12]: `ProposalStateType` and `IntegratorStateType` should be `namedtuple`s or `dataclass`es.  That change&mdash;along with some corresponding changes from indexing to field names&mdash;will make this code a lot easier to follow.

**Review by brandonwillard — CHANGES_REQUESTED** [review]

File: aehmc/metrics.py
Comment (brandonwillard) [inline aehmc/metrics.py:12]: This is the start of where the shape/broadcast information is being lost.

In your test, `inverse_mass_matrix.broadcastable == [True]`, which means that `inverse_mass_matrix.shape == (1,)`.  `aet.shape(inverse_mass_matrix)[0]` will not directly return one, though, so, when `shape` is used later as the `size` parameter for `norm_samples` below, the `RandomVariable` `Op` isn't able to determine that `norm_samples` is also broadcastable.  To the `RandomVariable` `Op`, `shape` could be anything, so it sets `norm_samples.broadcastable` to `[False]` just to be "safe".

I changed this line locally to use `aesara.tensor.shape.shape_tuple` instead, and that will make `shape` a `ScalarConstant` of `1`; however, it looks like there might be an issue with the `RandomVariable` `Op`, because `norm_samples.broadcastable` is still `[False]`.
`norm_samples` below will be given a shape

Comment (brandonwillard) [reply aehmc/metrics.py:12]:
```suggestion
    shape = shape_tuple(inverse_mass_matrix)[0]
```

File: aehmc/proposals.py
Comment (rlouf) [reply aehmc/proposals.py:12]: I agree but we're going to change something on the `aesara` side to be able to use them in `scan`.

Comment (brandonwillard) [reply aehmc/proposals.py:12]: I'm essentially waiting for us to replace the C backend with Numba as the default.  After that, we can very easily add new `Type`s for `tuple`s, `list`s, `str`s, `namedtuple`s, etc.  We could do that now, but those types would need to be in Python&mdash;and, thus, not as efficient to use&mdash;or C, and that would involve some very annoying and often involved CPython-via-string-templating work.

In the meantime, we can write our code to use `namedtuples` and then make the change when those types are available.

**Comment by rlouf** [comment]
I was able to track down the source of the bug. The trajectory is always interrupted early because the termination criterion returns `True` at the first step. This is related to what happens when the step index is 0, I'll investigate.

Other curious thing: after making the step index start from 0 instead of 1 I get the following error:

```        
if n_fixed_steps in [1, -1]:
    # We do not need to use the scan op anymore, so we can just return
    # the outputs and updates we have
     if condition is not None:
         _logger.warning(
>           (
             "When the number of steps is fixed and equal "
             "to 1, the provided stopping condition, {} is ignored",
             ).format(condition)
        )
E     AttributeError: 'tuple' object has no attribute 'format'

```

The error disappears when I delete the warning. I tried to reproduce this on a simpler example but could not.

**Comment by brandonwillard** [comment]
> I was able to track down the source of the bug. The trajectory is always interrupted early because the termination criterion returns `True` at the first step. This is related to what happens when the step index is 0, I'll investigate.
> 
> Other curious thing: after making the step index start from 0 instead of 1 I get the following error:
> 
> ```
> if n_fixed_steps in [1, -1]:
>     # We do not need to use the scan op anymore, so we can just return
>     # the outputs and updates we have
>      if condition is not None:
>          _logger.warning(
> >           (
>              "When the number of steps is fixed and equal "
>              "to 1, the provided stopping condition, {} is ignored",
>              ).format(condition)
>         )
> E     AttributeError: 'tuple' object has no attribute 'format'
> ```
> 
> The error disappears when I delete the call to `_logger.warning`. I tried to reproduce this on a simpler example but could not.

Where is this code coming from?

**Comment by rlouf** [comment]
It comes from the [implementation](https://github.com/aesara-devs/aesara/blob/15c64971dc5d033a8c4c2a425ebcb126d808fbd7/aesara/scan/basic.py#L763) of the `scan` function.

**Comment by brandonwillard** [comment]
Ah, yeah, that's a simple bug to fix.

**Comment by rlouf** [comment]
@brandonwillard @ricardoV94 the tests are failing because of the aesara bug mentioned above. It is otherwise good for review.

**Comment by brandonwillard** [comment]
> @brandonwillard @ricardoV94 the tests are failing because of the aesara bug mentioned above. It is otherwise good for review.

I'll create a new release of Aesara with that fix in just a minute.

**Review by brandonwillard — CHANGES_REQUESTED** [review]
If you change this to use Aesara >= 2.2.0, the `Scan` issue should be fixed.

File: aehmc/proposals.py
Comment (brandonwillard) [inline aehmc/proposals.py:19]: Is this missing a test?

File: aehmc/proposals.py
Comment (ricardoV94) [inline aehmc/proposals.py:155]: This was implemented twice. 

Also a reminder for me to check how this one performs compared with the more general default `logaddexp` rewrite we have. We can easily add this when there are only two terms being added, if it's any better

File: aehmc/proposals.py
Comment (rlouf) [reply aehmc/proposals.py:19]: I don't think this needs a unit test, and it should be covered by the integration tests.

Comment (brandonwillard) [reply aehmc/proposals.py:19]: Some form of coverage is needed, though.  

For instance, we want something that would alert us to any issues arising when the form of `state` changes, because that unpacking and/or the use of the unpacked items are an easy source of refactoring problems.

Comment (rlouf) [reply aehmc/proposals.py]: Turns out this function was not needed, so I removed it.

File: aehmc/proposals.py
Comment (rlouf) [reply aehmc/proposals.py:150]: Yes we should definitely check against the rewrite. In the meantime I am not so bothered by the duplication.

**Comment by rlouf** [comment]
The tests are now passing locally, but it looks like there is a problem with the `aesara` installation in the CI. Any idea why?

File: setup.py
Comment (brandonwillard) [inline setup.py:20]: This was removed because it's not being directly used by this library and it causes `pip` to overwrite the Conda environment's NumPy, which can cause the issue you're seeing in CI.

**Review by brandonwillard — CHANGES_REQUESTED** [review]

**Comment by rlouf** [comment]
Will add a logistic regression example to the notebook once #19 is merged.

**Review by brandonwillard — CHANGES_REQUESTED** [review]
Let's update those commit messages so that they follow [these guidelines](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html) from [`PULL_REQUEST_TEMPLATE.md`](https://github.com/aesara-devs/aehmc/blob/21b460dab7e9a47ac9d6f1ff792cd8a71fe26017/.github/PULL_REQUEST_TEMPLATE.md) and merge this.

**Review by brandonwillard — APPROVED** [review]

**Pull request merged**
